File: tpf2_mp_1/res/scripts/tpf2_mp/calendar_model.lua
local util = require "tpf2_mp/util"

local M = {}

M.SCHEMA_VERSION = 1
M.DEFAULT_MILLIS_PER_DAY = 2000
M.MIN_YEAR = 1400
M.MAX_YEAR = 9999

local function exactInteger(value)
  local number = tonumber(value)
  if not number or number ~= math.floor(number) then return nil end
  return number
end

local function leapYear(year)
  return year % 4 == 0 and (year % 100 ~= 0 or year % 400 == 0)
end

local function monthDays(year, month)
  local days = { 31, leapYear(year) and 29 or 28, 31, 30, 31, 30,
    31, 31, 30, 31, 30, 31 }
  return days[month]
end

function M.normaliseDate(value)
  if type(value) ~= "table" then return nil, "calendar date must be an object" end
  local year, month, day = exactInteger(value.year), exactInteger(value.month), exactInteger(value.day)
  if not year or year < M.MIN_YEAR or year > M.MAX_YEAR then
    return nil, "calendar year is outside the supported native range"
  end
  if not month or month < 1 or month > 12 then return nil, "calendar month is invalid" end
  local maximum = monthDays(year, month)
  if not day or day < 1 or day > maximum then return nil, "calendar day is invalid" end
  return { year = year, month = month, day = day }
end

local function daysFromCivil(date)
  local year, month, day = date.year, date.month, date.day
  if month <= 2 then year = year - 1 end
  local era = math.floor(year / 400)
  local yearOfEra = year - era * 400
  local monthPrime = month + (month > 2 and -3 or 9)
  local dayOfYear = math.floor((153 * monthPrime + 2) / 5) + day - 1
  local dayOfEra = yearOfEra * 365 + math.floor(yearOfEra / 4)
    - math.floor(yearOfEra / 100) + dayOfYear
  return era * 146097 + dayOfEra
end

local function civilFromDays(days)
  local era = math.floor(days / 146097)
  local dayOfEra = days - era * 146097
  local yearOfEra = math.floor((dayOfEra - math.floor(dayOfEra / 1460)
    + math.floor(dayOfEra / 36524) - math.floor(dayOfEra / 146096)) / 365)
  local year = yearOfEra + era * 400
  local dayOfYear = dayOfEra - (365 * yearOfEra + math.floor(yearOfEra / 4)
    - math.floor(yearOfEra / 100))
  local monthPrime = math.floor((5 * dayOfYear + 2) / 153)
  local day = dayOfYear - math.floor((153 * monthPrime + 2) / 5) + 1
  local month = monthPrime + (monthPrime < 10 and 3 or -9)
  if month <= 2 then year = year + 1 end
  return { year = year, month = month, day = day }
end

function M.addDays(date, delta)
  local valid, dateError = M.normaliseDate(date)
  if not valid then return nil, dateError end
  delta = exactInteger(delta)
  if not delta then return nil, "calendar day delta must be an integer" end
  local result = civilFromDays(daysFromCivil(valid) + delta)
  if result.year < M.MIN_YEAR or result.year > M.MAX_YEAR then
    return nil, "calendar advancement exceeds the supported native range"
  end
  return result
end

function M.newState()
  return {
    schemaVersion = M.SCHEMA_VERSION,
    managed = false,
    initialized = false,
    millisPerDay = M.DEFAULT_MILLIS_PER_DAY,
    startDate = nil,
    currentDate = nil,
    elapsedDays = 0,
    residualMillis = 0,
    lastEpoch = 0,
  }
end

function M.initialise(rules, networkMode)
  rules = type(rules) == "table" and rules or {}
  local date, dateError = M.normaliseDate(rules.calendarStartDate)
  if not date then return nil, dateError end
  local millis = exactInteger(rules.calendarMillisPerDay)
  if not millis or millis < 0 or millis > 86400000 then
    return nil, "calendar milliseconds per day are invalid"
  end
  return {
    schemaVersion = M.SCHEMA_VERSION,
    managed = networkMode == "network",
    initialized = true,
    millisPerDay = millis,
    startDate = util.deepCopy(date),
    currentDate = util.deepCopy(date),
    elapsedDays = 0,
    residualMillis = 0,
    lastEpoch = 0,
  }
end

function M.prepareSettlement(current, economyEpoch, scheduled, epochSeconds)
  current = M.migrate(current)
  if current.initialized ~= true then return nil, nil, "authored calendar is not initialised" end
  local epoch = exactInteger(economyEpoch)
  if not epoch or epoch ~= current.lastEpoch + 1 then
    return nil, nil, "calendar settlement is not the next economy epoch"
  end
  local seconds = exactInteger(epochSeconds)
  if not seconds or seconds < 1 or seconds > 86400 then
    return nil, nil, "calendar settlement duration is invalid"
  end
  local candidate = util.deepCopy(current)
  local advancedDays = 0
  if candidate.managed and scheduled == true and candidate.millisPerDay > 0 then
    local accumulated = candidate.residualMillis + seconds * 1000
    advancedDays = math.floor(accumulated / candidate.millisPerDay)
    candidate.residualMillis = accumulated % candidate.millisPerDay
    if advancedDays > 0 then
      local advanced, dateError = M.addDays(candidate.currentDate, advancedDays)
      if not advanced then return nil, nil, dateError end
      candidate.currentDate = advanced
      candidate.elapsedDays = candidate.elapsedDays + advancedDays
    end
  end
  candidate.lastEpoch = epoch
  local payload = {
    schemaVersion = M.SCHEMA_VERSION,
    economyEpoch = epoch,
    advanced = candidate.managed and scheduled == true,
    advancedDays = advancedDays,
    elapsedDays = candidate.elapsedDays,
    residualMillis = candidate.residualMillis,
    date = util.deepCopy(candidate.currentDate),
  }
  return candidate, payload
end

local function sameDate(left, right)
  return type(left) == "table" and type(right) == "table"
    and left.year == right.year and left.month == right.month and left.day == right.day
end

function M.verifySettlement(current, action, epochSeconds)
  if type(action) ~= "table" or type(action.calendar) ~= "table"
      or type(action.results) ~= "table" then
    return nil, "economy settlement requires an authored calendar payload"
  end
  local candidate, expected, prepareError = M.prepareSettlement(
    current, action.results.epoch, action.scheduled == true, epochSeconds)
  if not candidate then return nil, prepareError end
  local received = action.calendar
  local date, dateError = M.normaliseDate(received.date)
  if not date then return nil, dateError end
  if exactInteger(received.schemaVersion) ~= M.SCHEMA_VERSION
    or exactInteger(received.economyEpoch) ~= expected.economyEpoch
    or type(received.advanced) ~= "boolean" or received.advanced ~= expected.advanced
    or exactInteger(received.advancedDays) ~= expected.advancedDays
    or exactInteger(received.elapsedDays) ~= expected.elapsedDays
    or exactInteger(received.residualMillis) ~= expected.residualMillis
    or not sameDate(date, expected.date) then
    return nil, "economy settlement calendar payload diverges from deterministic replay"
  end
  return candidate, expected
end

function M.migrate(value, fallbackDate, fallbackEpoch, fallbackMillis)
  local result = M.newState()
  if type(value) == "table" then
    result.managed = value.managed == true
    result.initialized = value.initialized == true
    result.millisPerDay = util.clamp(util.integer(
      value.millisPerDay, M.DEFAULT_MILLIS_PER_DAY), 0, 86400000)
    result.elapsedDays = math.max(0, util.integer(value.elapsedDays, 0))
    result.residualMillis = math.max(0, util.integer(value.residualMillis, 0))
    if result.millisPerDay > 0 then result.residualMillis = result.residualMillis % result.millisPerDay
    else result.residualMillis = 0 end
    result.lastEpoch = math.max(0, util.integer(value.lastEpoch, 0))
    result.startDate = M.normaliseDate(value.startDate)
    result.currentDate = M.normaliseDate(value.currentDate)
    if not result.startDate then result.startDate = util.deepCopy(result.currentDate) end
    if result.initialized and not result.currentDate then result.initialized = false end
  end
  if not result.initialized then
    local observed = M.normaliseDate(fallbackDate)
    if observed then
      result.initialized = true
      result.managed = true
      result.millisPerDay = util.clamp(util.integer(
        fallbackMillis, M.DEFAULT_MILLIS_PER_DAY), 0, 86400000)
      result.startDate = util.deepCopy(observed)
      result.currentDate = util.deepCopy(observed)
      result.elapsedDays = 0
      result.residualMillis = 0
      result.lastEpoch = math.max(0, util.integer(fallbackEpoch, 0))
    end
  end
  result.schemaVersion = M.SCHEMA_VERSION
  return result
end

function M.digestView(value)
  local current = M.migrate(value)
  return {
    schemaVersion = M.SCHEMA_VERSION,
    managed = current.managed,
    initialized = current.initialized,
    millisPerDay = current.millisPerDay,
    startDate = util.deepCopy(current.startDate),
    currentDate = util.deepCopy(current.currentDate),
    elapsedDays = current.elapsedDays,
    residualMillis = current.residualMillis,
    lastEpoch = current.lastEpoch,
  }
end

return M

File: companion/tpf2mp/calendar_model.py
from __future__ import annotations

import copy
from datetime import date, timedelta
from typing import Any, Mapping

from .protocol import ProtocolError


SCHEMA_VERSION = 1
DEFAULT_MILLIS_PER_DAY = 2000
MIN_YEAR = 1400
MAX_YEAR = 9999


def _integer(value: Any, name: str, minimum: int, maximum: int) -> int:
    if isinstance(value, bool) or not isinstance(value, int) or not minimum <= value <= maximum:
        raise ProtocolError(f"{name} is invalid")
    return value


def normalize_date(value: Any) -> dict[str, int]:
    if not isinstance(value, Mapping) or set(value) != {"year", "month", "day"}:
        raise ProtocolError("calendar date must contain year, month, and day")
    year = _integer(value.get("year"), "calendar year", MIN_YEAR, MAX_YEAR)
    month = _integer(value.get("month"), "calendar month", 1, 12)
    day_value = _integer(value.get("day"), "calendar day", 1, 31)
    try:
        native = date(year, month, day_value)
    except ValueError as exc:
        raise ProtocolError("calendar date is invalid") from exc
    return {"year": native.year, "month": native.month, "day": native.day}


def add_days(value: Any, delta: int) -> dict[str, int]:
    current = normalize_date(value)
    _integer(delta, "calendar day delta", 0, 10_000_000)
    try:
        advanced = date(current["year"], current["month"], current["day"]) + timedelta(days=delta)
    except (OverflowError, ValueError) as exc:
        raise ProtocolError("calendar advancement exceeds the supported native range") from exc
    if not MIN_YEAR <= advanced.year <= MAX_YEAR:
        raise ProtocolError("calendar advancement exceeds the supported native range")
    return {"year": advanced.year, "month": advanced.month, "day": advanced.day}


def new_state() -> dict[str, Any]:
    return {
        "schemaVersion": SCHEMA_VERSION,
        "managed": False,
        "initialized": False,
        "millisPerDay": DEFAULT_MILLIS_PER_DAY,
        "startDate": None,
        "currentDate": None,
        "elapsedDays": 0,
        "residualMillis": 0,
        "lastEpoch": 0,
    }


def migrate(value: Any) -> dict[str, Any]:
    result = new_state()
    if not isinstance(value, Mapping):
        return result
    result["managed"] = value.get("managed") is True
    result["initialized"] = value.get("initialized") is True
    result["millisPerDay"] = _integer(
        value.get("millisPerDay", DEFAULT_MILLIS_PER_DAY),
        "calendar milliseconds per day", 0, 86_400_000,
    )
    result["elapsedDays"] = _integer(value.get("elapsedDays", 0), "calendar elapsed days", 0, 10_000_000)
    result["residualMillis"] = _integer(
        value.get("residualMillis", 0), "calendar residual milliseconds", 0, 86_400_000,
    )
    result["lastEpoch"] = _integer(value.get("lastEpoch", 0), "calendar last epoch", 0, 1_000_000_000)
    if result["millisPerDay"] > 0:
        if result["residualMillis"] >= result["millisPerDay"]:
            raise ProtocolError("calendar residual milliseconds exceed one native day")
    elif result["residualMillis"] != 0:
        raise ProtocolError("paused calendar cannot carry a residual")
    if value.get("startDate") is not None:
        result["startDate"] = normalize_date(value["startDate"])
    if value.get("currentDate") is not None:
        result["currentDate"] = normalize_date(value["currentDate"])
    if result["initialized"] and (result["startDate"] is None or result["currentDate"] is None):
        raise ProtocolError("initialized calendar is missing a date")
    return result


def prepare_settlement(
    current_value: Any,
    economy_epoch: int,
    scheduled: bool,
    epoch_seconds: int,
) -> tuple[dict[str, Any], dict[str, Any]]:
    current = migrate(current_value)
    if current["initialized"] is not True:
        raise ProtocolError("authored calendar is not initialised")
    epoch = _integer(economy_epoch, "calendar economy epoch", 1, 1_000_000_000)
    if epoch != current["lastEpoch"] + 1:
        raise ProtocolError("calendar settlement is not the next economy epoch")
    seconds = _integer(epoch_seconds, "calendar epoch duration", 1, 86_400)
    candidate = copy.deepcopy(current)
    advanced_days = 0
    if candidate["managed"] and scheduled and candidate["millisPerDay"] > 0:
        accumulated = candidate["residualMillis"] + seconds * 1000
        advanced_days, candidate["residualMillis"] = divmod(accumulated, candidate["millisPerDay"])
        if advanced_days:
            candidate["currentDate"] = add_days(candidate["currentDate"], advanced_days)
            candidate["elapsedDays"] += advanced_days
    candidate["lastEpoch"] = epoch
    payload = {
        "schemaVersion": SCHEMA_VERSION,
        "economyEpoch": epoch,
        "advanced": candidate["managed"] and scheduled,
        "advancedDays": advanced_days,
        "elapsedDays": candidate["elapsedDays"],
        "residualMillis": candidate["residualMillis"],
        "date": copy.deepcopy(candidate["currentDate"]),
    }
    return candidate, payload


def validate_payload(value: Any) -> dict[str, Any]:
    if not isinstance(value, Mapping) or set(value) != {
        "schemaVersion", "economyEpoch", "advanced", "advancedDays",
        "elapsedDays", "residualMillis", "date",
    }:
        raise ProtocolError("economy.settle calendar payload is malformed")
    if value.get("schemaVersion") != SCHEMA_VERSION:
        raise ProtocolError("economy.settle calendar schema is unsupported")
    if not isinstance(value.get("advanced"), bool):
        raise ProtocolError("economy.settle calendar advanced flag is invalid")
    result = {
        "schemaVersion": SCHEMA_VERSION,
        "economyEpoch": _integer(value.get("economyEpoch"), "calendar economy epoch", 1, 1_000_000_000),
        "advanced": value["advanced"],
        "advancedDays": _integer(value.get("advancedDays"), "calendar advanced days", 0, 10_000_000),
        "elapsedDays": _integer(value.get("elapsedDays"), "calendar elapsed days", 0, 10_000_000),
        "residualMillis": _integer(value.get("residualMillis"), "calendar residual milliseconds", 0, 86_400_000),
        "date": normalize_date(value.get("date")),
    }
    return result


def validate_match_rules(rules: Mapping[str, Any]) -> None:
    if "calendarMillisPerDay" in rules:
        _integer(rules.get("calendarMillisPerDay"),
                 "match.initialise calendarMillisPerDay", 0, 86_400_000)
    if "calendarStartDate" in rules:
        normalize_date(rules.get("calendarStartDate"))


def apply_settlement(model: dict[str, Any], action: Mapping[str, Any], economy_epoch: int, epoch_seconds: int) -> None:
    received = validate_payload(action.get("calendar"))
    candidate, expected = prepare_settlement(
        model.get("calendar"), economy_epoch, action.get("scheduled") is True, epoch_seconds,
    )
    if received != expected:
        raise ProtocolError("economy.settle calendar payload diverges from deterministic replay")
    model["calendar"] = candidate
